core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3523,15 +3523,30 @@ object Stream extends StreamLowPriority {
     * All elements that are available, up to the specified limit,
     * are dequeued and emitted as a single chunk.
     */
-  def fromQueueUnterminated[F[_]: Functor, A](
+  def fromQueueUnterminated[F[_], A](
       queue: QueueSource[F, A],
       limit: Int = Int.MaxValue
-  ): Stream[F, A] =
-    fromQueueNoneTerminatedSingletons_[F, A](
-      queue.take.map(a => Some(a)),
-      queue.tryTake.map(_.map(a => Some(a))),
-      limit
-    )
+  )(implicit F: Functor[F]): Stream[F, A] =
+    F match {
+      case f0: Monad[F] =>
+        val someLimit = Some(limit)
+
+        /** use non-blocking tryTakeN, which is possibly more performant than n * take */
+        Stream
+          .eval(queue.tryTakeN(someLimit)(f0))
+          .flatMap {
+            case Nil => Stream.eval(queue.take)
+            case as  => Stream.emits(as)
+          }
+          .repeat
```

Comment:
Sorry, now that I realize we do need `Monad[F]` for `tryTakeN`, actually it's more efficient to use `flatMap` on `F[_]` than `Stream`. So basically what you originally proposed, sorry.

We also need to add one more feature to preserve the original semantics: after a `take` we should attempt to `tryTakeN` to get any other elements that are immediately available.

```suggestion
        /** use non-blocking tryTakeN, which is possibly more performant than n * take */
        if (limit > 1) {
          val someLimit = Some(limit)
          val someLimitLess1 = Some(limit - 1)
          Stream.evalSeq {
            queue.tryTakeN(someLimit)(f0).flatMap {
              case Nil => queue.take.map2(queue.tryTakeN(someLimitLess1)(f0)))(_ :: _)
              case as => F.pure(as)
            }
          }.repeat
        } else Stream.repeatEval(queue.take)
```

Reply:
`map2` goes through an extra step of tupling, please review the direct version with `flatMap` + `map`.
Also, we can still do `tryTakeN` even for `n == 1`.

Reply:
> `map2` goes through an extra step of tupling

where is the tupling?

> Also, we can still do `tryTakeN` even for `n == 1`.

What is the point of `tryTakeN` for `n == 1`? It's just adding an extra and unnecessary step. When you `take` you will either get the element right away or have to wait. What is `tryTakeN` bringing to the table here?

Reply:
> > `map2` goes through an extra step of tupling
> 
> where is the tupling?

```scala
  def map2[A, B, Z](fa: F[A], fb: F[B])(f: (A, B) => Z): F[Z] =
    map(product(fa, fb))(f.tupled)
```

> 
> > Also, we can still do `tryTakeN` even for `n == 1`.
> 
> What is the point of `tryTakeN` for `n == 1`? It's just adding an extra and unnecessary step. When you `take` you will either get the element right away or have to wait. What is `tryTakeN` bringing to the table here?

I looked at the internals of a generic `take` vs `tryTakeN` for the async queue and got the notion that the latter might be actually more performant. If I am wrong, we shall revert to your proposition.

Reply:
1. That's not the final implementation used. Please see the override in `FlatMap` which does not tuple.

https://github.com/typelevel/cats/blob/7e04753b20a21f6f0e35eb2e6db1a0fb4ae512d2/core/src/main/scala/cats/FlatMap.scala#L113-L114

2. Hmm, I'm getting that notion as well. But I'm not sure why it's that way and I think we can fix it in Cats Effect.

Reply:
> I think we can fix it in Cats Effect.

Nope, it's needed for correctness. I tried moving the `uncancelable` inside the `catch` but that allows an element to be lost on cancelation.

https://github.com/typelevel/cats-effect/blob/25c948b065de0c7c77c9655162cdeb093f0b2b3b/std/shared/src/main/scala/cats/effect/std/Queue.scala#L641-L643

Ok, so for the record that `uncancelable` in `take` is adding some allocation overhead over `tryTakeN`. However I still think the simpler implementation in terms of `take` is better here. At the end of the day limiting the chunk size `n == 1` is sub-optimal and I would expect the per-element overhead of `Stream` machinery to dwarf the `uncancelable`. I'd be surprised if a benchmark measures a difference.

Reply:
Please see the revised version.